Source/Dafny/Resolver.cs

```diff
@@ -10042,6 +10068,41 @@ public class ResolveTypeReturn
       ConstrainTypeExprBool(s.Expr, "type of RHS of assign-such-that statement must be boolean (got {0})");
     }
 
+    // TODO is there already a library of convenience AST construction methods like this one?
+    private Expression VarDotMethod(IToken tok, string varname, string methodname) {
+      return new ApplySuffix(tok, new ExprDotName(tok, new IdentifierExpr(tok, varname), methodname, null), new List<Expression>() { });
+    }
+
+    // desugars "y :- MethodOrExpression" into 
+    // "var temp := MethodOrExpression; if temp.IsFailure() { return temp.PropagateFailure(); } y := temp.Extract();"
+    private Statement DesugarAssignOrReturnStmt(AssignOrReturnStmt s, ICodeContext codeContext) {
+      // TODO how to mark this as autogenerated? Should we create an equivalent of AutoGeneratedExpression for statements?
```

Comment:
Since this code is going to go into a `ResolveStmts` field, I think it's okay not to further mark it as auto-generated.